File: src/runner.ts
import { Check, CheckContext, CheckResult } from './types.js';

export interface RunnerOptions {
  checks: Check[];
  context: CheckContext;
  onCheckStart?: (check: Check) => void;
  onCheckComplete?: (check: Check, result: CheckResult) => void;
}

export interface RunnerResult {
  results: CheckResult[];
  totalDuration: number;
  passed: number;
  failed: number;
}

export async function runChecks(options: RunnerOptions): Promise<RunnerResult> {
  const { checks, context, onCheckStart, onCheckComplete } = options;
  const results: CheckResult[] = [];
  const startTime = performance.now();

  for (const check of checks) {
    onCheckStart?.(check);

    try {
      const timeoutPromise = new Promise<CheckResult>((_, reject) => {
        setTimeout(() => {
          reject(new Error(`Check timed out after ${context.timeout}ms`));
        }, context.timeout);
      });

      const result = await Promise.race([
        check.run(context),
        timeoutPromise,
      ]);

      results.push(result);
      onCheckComplete?.(check, result);
    } catch (error) {
      const result: CheckResult = {
        name: check.name,
        success: false,
        duration: 0,
        message: (error as Error).message || 'Unknown error',
        details: 'Check threw an exception',
      };
      results.push(result);
      onCheckComplete?.(check, result);
    }
  }

  const totalDuration = performance.now() - startTime;
  const passed = results.filter(r => r.success).length;
  const failed = results.filter(r => !r.success).length;

  return {
    results,
    totalDuration,
    passed,
    failed,
  };
}

File: python/runner.py
"""Sequential check runner matching TypeScript version."""
import time
import asyncio
from typing import List
from .types import Check, CheckContext, CheckResult
from .reporter import Reporter


async def run_check_with_timeout(
    check: Check,
    context: CheckContext,
    timeout_ms: int
) -> CheckResult:
    """Run a single check with timeout protection.

    Args:
        check: The check to run
        context: Shared check context
        timeout_ms: Timeout in milliseconds

    Returns:
        CheckResult from the check or timeout error
    """
    try:
        result = await asyncio.wait_for(
            check.run(context),
            timeout=timeout_ms / 1000.0
        )
        return result
    except asyncio.TimeoutError:
        return CheckResult(
            name=check.name,
            success=False,
            duration=timeout_ms,
            message=f"Check timed out after {timeout_ms}ms",
            suggestion="Increase timeout or check network connectivity"
        )


async def run_checks(
    checks: List[Check],
    context: CheckContext,
    reporter: Reporter
) -> List[CheckResult]:
    """Run checks sequentially, passing context between them.

    Args:
        checks: List of checks to run
        context: Shared check context
        reporter: Reporter for output

    Returns:
        List of CheckResults from all checks
    """
    results: List[CheckResult] = []

    reporter.start()

    for check in checks:
        reporter.on_check_start(check)

        start_time = time.perf_counter()
        try:
            result = await run_check_with_timeout(
                check,
                context,
                context.timeout
            )
            # Update duration with actual time if not set by timeout
            if result.duration == 0:
                result.duration = (time.perf_counter() - start_time) * 1000
        except Exception as e:
            result = CheckResult(
                name=check.name,
                success=False,
                duration=(time.perf_counter() - start_time) * 1000,
                message=f"Check failed with exception: {e}",
                suggestion="Check logs for details"
            )

        results.append(result)
        reporter.on_check_complete(result)

    reporter.finish(results)
    return results
